### PR TITLE
Add support for nested generic types

### LINKED ISSUE #1694: Length cannot be less than zero
Hello, with this benchmark
```csharp
public class Test
{
    [Benchmark]
    public Dictionary<string, int>.Enumerator ConcreteGetEnumerator() => default;
}
```

I get a
```
// Validating benchmarks:
// ***** BenchmarkRunner: Start   *****
// ***** Found 1 benchmark(s) in total *****
// ***** Building 1 exe(s) in Parallel: Start   *****
// ***** Done, took 00:00:00 (0.07 sec)   *****
// Found 1 benchmarks:
//   Test.ConcreteGetEnumerator: DefaultJob

// Generate Exception: Length cannot be less than zero. (Parameter 'length')
```

BenchmarkDotNet: 0.12.1
Target Framework: .NET Core 2.1 (also tested with 3.1)

If you can't reproduce on your side I'll dig deeper.

### PR DESCRIPTION
**Pull request by YegorStepanov**
Fix #1694

### CODE REVIEW

**Comment by timcassell** [comment]
Maybe I'm missing something, but I don't see how the change fixes types nested under generic types. Shouldn't the change be in the `while (child.IsNested && parent != null)` loop and include the `<{args}>`?

Does it also work for these?

```cs
public class Type1<T>
{
    public class Type2<U1, U2>
    {
        public class Type3<V1, V2, V3>
        {
            public class Type4
            {
            
            }
            public class Type4<W>
            {
            
            }
        }
    }
}
```

**Comment by YegorStepanov** [comment]
Hi @timcassell 

```C#
public class Generic1<T1> //int
{                                    
    public class NonGeneric2          
    {                                
        public class Generic3<T2> { } //float
    }                                
}                                    

// BenchmarkDotNet.Tests.ReflectionTests.Generic1`1.NonGeneric2.Generic3<System.Int32, System.Single>
```

```C#
public class NonGeneric1                  
{                                        
    public class Generic2<T> //int
    {                                    
        public class NonGeneric3          
        {                                
            public class Generic4<T1> { } //float
        }                                
    }                                    
}

// BenchmarkDotNet.Tests.ReflectionTests.NonGeneric1.Generic2`1.NonGeneric3.Generic4<System.Int32, System.Single>                                        
```

```C#
public class Type1<T> //int
{                                     
    public class Type2                 
    {                                 
        public class Type3<V1, V2, V3> //float, double, long
        {                             
            public class Type4 { }     
                                      
            public class Type4<W> { }  //short
        }                             
    }                                 
}

// BenchmarkDotNet.Tests.ReflectionTests.Type1`1.Type2.Type3`3.Type4

// BenchmarkDotNet.Tests.ReflectionTests.Type1`1.Type2.Type3`3.Type4<System.Int32, System.Single, System.Double, System.Int64, System.Int16>
```

Is it correct?

I'll add your test.

**Comment by YegorStepanov** [comment]
I'll try to compile real benchmarks now

**Comment by timcassell** [comment]
@YegorStepanov You got the C# type names wrong.

```
BenchmarkDotNet.Tests.ReflectionTests.Generic1`1.NonGeneric2.Generic3<System.Int32, System.Single>
```

Should be

```
BenchmarkDotNet.Tests.ReflectionTests.Generic1<System.Int32>.NonGeneric2.Generic3<System.Single>
```

**Comment by YegorStepanov** [comment]
`type.GetGenericArguments()` returns all generic parameters. The type parameters are erased on `.DeclaredType`:

image

Example:
```C#
public class Type1<T1, T2>
{
    public class Type2<V1, V2> { }
}
```
For `type = typeof(Type1<int, long>.Type2<float, double>)`:
1) Call `type.GetGenericArguments()` and add them to the stack:
```C#
Type[4] { [int], [long], [float], [double] }

Stack<Type>(4) { [double], [float], [long], [int] }
```
2) type.Name
```C#
"Type2`2"
```
3) `stack.Pop()` twice
```C#
Type[2] { [double], [float] }
```
4) reverse it
```C#
Type[2] { [float], [double] }
```
5) add them to the name

6) type.Name
```C#
"Type1`2"
```
7) `stack.Pop()` twice
```C#
Type[2] { [long], [int] }
```
8) reverse it
```C#
Type[2] { [int], [long] }
```
9) add them to the name

**Review by timcassell — APPROVED** [review]

File: tests/BenchmarkDotNet.Tests/ReflectionTests.cs [hunk 1]
```diff
@@ -26,6 +26,7 @@ public void GetCorrectCSharpTypeNameReturnsCSharpFriendlyTypeName()
             CheckCorrectTypeName("BenchmarkDotNet.Tests.ReflectionTests.NestedNonGeneric1.NestedNonGeneric2", typeof(NestedNonGeneric1.NestedNonGeneric2));
             CheckCorrectTypeName("BenchmarkDotNet.Tests.ReflectionTests.NestedNonGeneric1.NestedGeneric2<System.Int16, System.Boolean, System.Decimal>",
                 typeof(NestedNonGeneric1.NestedGeneric2<short, bool, decimal>));
+            CheckCorrectTypeName("BenchmarkDotNet.Tests.ReflectionTests.NestedGeneric1`1.NestedNonGeneric2", typeof(NestedGeneric1<int>.NestedNonGeneric2));
```
[the remark this answers is not in the document]
Comment (YegorStepanov) [reply tests/BenchmarkDotNet.Tests/ReflectionTests.cs]: Hmm, I have force pushed the branch to delete my test from yesterday. 

How were you able to find it? :)

[the remark this answers is not in the document]
Comment (timcassell) [reply tests/BenchmarkDotNet.Tests/ReflectionTests.cs]: I wrote that before you pushed your changes.

**Review by adamsitnik — APPROVED** [review]
@YegorStepanov thank you for providing the fix

@timcassell thank you for reviewing the PR!

LGTM, let's merge it and close the bug.

**Review by adamsitnik — APPROVED** [review]
@YegorStepanov thank you for providing the fix

@timcassell thank you for reviewing the PR!

LGTM, let's merge it and close the bug.

File: src/BenchmarkDotNet/Extensions/ReflectionExtensions.cs [hunk 2]
```diff
@@ -51,31 +51,43 @@ internal static string GetCorrectCSharpTypeName(this Type type, bool includeName
...
+        private static IEnumerable<string> GetNestedTypeNames(Type type, bool includeGenericArgumentsNamespace)
+        {
+            var allTypeParameters = new Stack<Type>(type.GetGenericArguments());
+
+            Type currentType = type;
+            while (currentType != null)
+            {
+                string name = currentType.Name.Replace("&", string.Empty);
+
+                if (name.Contains('`'))
```
[the remark this answers is not in the document]
Comment (YegorStepanov) [reply src/BenchmarkDotNet/Extensions/ReflectionExtensions.cs:73]: Unfortunately yes. I tried to find a better solution, but couldn't.

**Pull request merged**